File: problems/src/Problem18.java
import java.io.*;
import java.util.ArrayList;

/**
 * Created by YSingh on 08/02/17.
 */
public class Problem18 {
    public static void main(String[] args) throws IOException {
        FileInputStream fis = new FileInputStream("problems/src/Problem18");
        DataInputStream in = new DataInputStream(fis);
        BufferedReader br = new BufferedReader(new InputStreamReader(in));
        String line = null;
        int T = 15;
        int row = 15;
        int col = 15;

        int [][] arr = new int[row + 1][col + 1];
        String[] str;

        int i = 0;
        while (T > 0) {
            i++;
            T--;
            line = br.readLine();
            str = line.split(" ");
            for (int j = 0; j < str.length; j++) {
                arr[i][j + 1] = Integer.parseInt(str[j]);
            }
        }
        int sum = 0;
        System.out.println(getMaxSum(arr, 1, 1, row, col, sum));
        br.close();
    }

    private static int getMaxSum(int[][] arr, int i, int j, int row, int col, int sum) {
        if(i > row) {
            return sum;
        }

        int left = getMaxSum(arr, i + 1, j, row, col, sum + arr[i][j]);
        int right = getMaxSum(arr, i + 1, j + 1, row, col, sum + arr[i][j]);

        return Math.max(left, right);
    }
}

File: js/problem18.js
const str = `75
95 64
17 47 82
18 35 87 10
20 04 82 47 65
19 01 23 75 03 34
88 02 77 73 07 63 67
99 65 04 28 06 16 70 92
41 41 26 56 83 40 80 70 33
41 48 72 33 47 32 37 16 94 29
53 71 44 65 25 43 91 52 97 51 14
70 11 33 28 77 73 17 78 39 68 17 57
91 71 52 38 17 14 91 43 58 50 27 29 48
63 66 04 68 89 53 67 30 73 16 69 87 40 31
04 62 98 27 23 09 70 98 73 93 38 53 60 04 23`;

const rows = str.split('\n');
const matrix = new Array(rows.length);

for(let i = 0; i < rows.length; i++) {
	let cols = rows[i].split(' ');
	matrix[i] = new Array(cols.length);
	for(let j = 0; j < cols.length; j++) {
		matrix[i][j] = cols[j];
	}
}

function getMaxSum(matrix, i, j) {
	if(matrix[i] === undefined || matrix[i][j] === undefined) {
		return 0;
	}
	return parseInt(matrix[i][j]) + Math.max(getMaxSum(matrix, i + 1, j), getMaxSum(matrix, i + 1, j + 1));
}

console.log(getMaxSum(matrix, 0, 0));